### PR TITLE
imaginary - fix generation of heif, pdf, illustrator and svg thumbnails

### LINKED ISSUE #36543: Setup Imaginary in order to allow it to Rasterise PDF Files for NC Previews
**Is your feature request related to a problem? Please describe.**
Current configuration of Imaginary as of NC 25.0.3 does not support rendering a preview image for PDF files. Imaginary is capable of doing this seemingly out of the box with Ubuntu 22.04 `libvips42` package (unsure of build for AIO docker but also tested); request to allow for use of Imaginary rather than having to use ImageMagick.

**Describe the solution you'd like**
`Imaginary.php` file in `/lib/private/Preview` requires editing to support mimetype `Application/PDF` and probably some form of logic in the operations section of function `getCroppedThumbnail` to perform the necessary `convert` operation to get a usable preview image for NC. I am not really a coder, so I'm not sure what other logic might be desirable in the operation section of the function for PDF previews. It's likely that you'd still want the other params defined for previews to also be applied to PDF rasterisation images.

**Describe alternatives you've considered**
An evaluation of capability is tested working both with an Imaginary source build on Ubuntu 22.04 and using the Nextcloud AIO Imaginary docker `:develop` branch image.

**Additional context**
See discussion [#1906](https://github.com/nextcloud/all-in-one/discussions/1906#discussioncomment-4866032) from NC AIO that demonstrates the capability.

### PR DESCRIPTION
**Pull request by szaimen**
This is best reviewed like this: https://github.com/nextcloud/server/pull/36546/files?diff=unified&w=1

Close https://github.com/nextcloud/server/issues/36543

### CODE REVIEW

File: lib/private/Preview/Imaginary.php [hunk 1]
```diff
@@ -57,7 +57,7 @@ public function getMimeType(): string {
 	}
 
 	public static function supportedMimeTypes(): string {
-		return '/image\/(bmp|x-bitmap|png|jpeg|gif|heic|svg|tiff|webp)/';
+		return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|heif|svg|webp)|application\/pdf)/';
```
Comment (szaimen) [inline lib/private/Preview/Imaginary.php:60]: https://github.com/nextcloud/server/blob/37bb33c5799b834dfef3fb73936bd0e5a4773fb8/resources/config/mimetypemapping.dist.json#L180

Comment (szaimen) [reply lib/private/Preview/Imaginary.php:60]:
```suggestion
		return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|heif|svg\+xml|webp)|application\/pdf)/';
```

File: lib/private/Preview/Imaginary.php [hunk 2]
```diff
@@ -81,18 +81,30 @@ public function getCroppedThumbnail(File $file, int $maxX, int $maxY, bool $crop
...
 		$httpClient = $this->service->newClient();
 
+		$operations = [];
+
 		switch ($file->getMimeType()) {
 			case 'image/gif':
 			case 'image/png':
 				$mimeType = 'png';
 				break;
+			case 'image/svg':
```
Comment (szaimen) [inline lib/private/Preview/Imaginary.php:91]: Check if this is then really needed or if yes, add it in the case above (with +xml ending)

**Review by st3iny — APPROVED** [review]
Tested and works.

**Review by miaulalala — APPROVED** [review]

**Comment by solracsf** [comment]
@szaimen any objections backporting to 24?

**Comment by szaimen** [comment]
We can try

**Comment by enoch85** [comment]
@szaimen Does this require `OC\\Preview\\PDF` in `config.php`, or is `OC\\Preview\\Imaginary` enough?

**Comment by szaimen** [comment]
> or is `OC\\Preview\\Imaginary` enough?

this is enough

**Comment by solracsf** [comment]
@szaimen maybe document somewhere that, once `OC\Preview\Imaginary` is defined, [all these mimetypes are automatically supported](https://github.com/nextcloud/server/blob/3a56ecc8cb6556372d7615679f0ec84bc95e8af5/lib/private/Preview/Imaginary.php#L60) and so no need to define them?

**Pull request merged**